src/ott/neural/solvers/expectile_neural_dual.py

```diff
@@ -0,0 +1,655 @@
+        (self.state_g, self.state_f, loss, loss_f, loss_g,
+         w_dist) = self.train_step(self.state_g, self.state_f, train_batch)
+
+      if self.logging and step % self.log_freq == 0:
+        self._update_logs(train_logs, loss_f, loss_g, w_dist)
+
+      if callback is not None:
+        _ = callback(step, self.to_dual_potentials())
+
+      if step != 0 and step % self.valid_freq == 0:
+        valid_batch["source"] = jnp.asarray(next(validloader_source))
+        valid_batch["target"] = jnp.asarray(next(validloader_target))
+
+        valid_loss_f, valid_loss_g, valid_w_dist = self.valid_step(
+            self.state_f, self.state_g, valid_batch
+        )
+
+        if self.logging:
+          self._update_logs(
+              valid_logs, valid_loss_f, valid_loss_g, valid_w_dist
+          )
+
+    return {"train_logs": train_logs, "valid_logs": valid_logs}
+
+  def _get_train_step(
+      self
+  ) -> Callable[
+      [PotentialTrainState, PotentialTrainState, Dict[str, jnp.ndarray]], Tuple[
+          PotentialTrainState, PotentialTrainState, jnp.ndarray, jnp.ndarray,
+          jnp.ndarray, jnp.ndarray]]:
+    loss_fn = self._euclidean_loss_fn if self.use_dot_product else self._loss_fn
+
+    @jax.jit
+    def step_fn(state_f, state_g, batch):
+      grad_fn = jax.value_and_grad(loss_fn, argnums=[0, 1], has_aux=True)
+      (loss, (loss_f, loss_g, w_dist)), (grads_f, grads_g) = grad_fn(
+          state_f.params,
+          state_g.params,
+          state_f.potential_gradient_fn,
+          state_g.potential_value_fn,
+          batch,
+      )
+
+      return (
+          state_f.apply_gradients(grads=grads_f),
+          state_g.apply_gradients(grads=grads_g), loss, loss_f, loss_g, w_dist
+      )
+
+    return step_fn
+
+  def _get_valid_step(
+      self
+  ) -> Callable[
+      [PotentialTrainState, PotentialTrainState, Dict[str, jnp.ndarray]], Tuple[
+          jnp.ndarray, jnp.ndarray, jnp.ndarray]]:
+    loss_fn = self._euclidean_loss_fn if self.use_dot_product else self._loss_fn
+
+    @jax.jit
+    def step_fn(state_f, state_g, batch):
+      grad_fn = jax.value_and_grad(loss_fn, argnums=[0, 1], has_aux=True)
+      (loss, (loss_f, loss_g, w_dist)), _ = grad_fn(
+          state_f.params,
+          state_g.params,
+          state_f.potential_gradient_fn,
+          state_g.potential_value_fn,
+          batch,
+      )
+
+      return loss_f, loss_g, w_dist
+
+    return step_fn
+
+  def _loss_fn(
+      self, params_f: frozen_dict.FrozenDict[str, jnp.ndarray],
+      params_g: frozen_dict.FrozenDict[str, jnp.ndarray],
+      gradient_f: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                           PotentialGradientFn_t],
+      g_value: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                        PotentialValueFn_t], batch: Dict[str, jnp.ndarray]
+  ) -> Tuple[jnp.ndarray, Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray]]:
+
+    if self.is_bidirectional:
```

Comment:
Wouldn't it be possible to do instead:
```python
transport = self.to_dual_potentials().transport
```

Reply:
If not, please refactor using the twist operator.

Reply:
Replaced by `transport = ENOTPotentials(
        gradient_f(params_f), g_value(params_g), self.cost_fn,
        self.is_bidirectional, self.use_dot_product
    ).transport`